software/common/libs/framing/frame_detector.h

```diff
@@ -0,0 +1,213 @@
+        restart_rx();
+        break;
+      default:
+        state_ = State::Lost;
+        restart_rx();
+        break;
+    }
+  }
+
+  // Callback method called when RxBuffer receives a marker character;
+  // Note: MARK character is already written to RxBuffer so all conditions
+  // evaluating rx_buffer_.received_length() take that into account
+  void on_character_match() override {
+    marker_count_++;
+
+    if (data_received()) {
+      on_char_match_with_data();
+    } else if (only_markers_received()) {
+      on_char_match_with_markers_only();
+    } else {
+      // Should never end up here
+      // DMA is not working?
+      // TODO alert, safe reset
+      restart_rx();
+    }
+  }
+
+  // Callback method called when underlying Rx system experiences an error
+  void on_rx_error([[maybe_unused]] RxError e) override {
+    switch (state_) {
+      case State::Lost:
+      case State::WaitForStartMarker:
+        // no change
+        break;
+      case State::ReceivingFrame:
+        state_ = State::Lost;
+        restart_rx();
+        break;
+    }
+  }
+
+  // Callback method called when RxBuffer is full
+  void on_rx_complete() override {
+    // We should never reach the full read of rx buffer.
+    // If we get here, this means there are no marker chars in the stream, so we are lost.
+    state_ = State::Lost;
+    restart_rx();
+  }
+
+  // Returns the last successfully detected frame, resets the frame_available flag.
+  // If frame_available was returning true before this call, it will return false afterwards
+  uint8_t *take_frame() {
+    frame_available_ = false;
+    return frame_buffer_;
+  }
+
+  // Returns the length of the last successfully detected frame.
+  uint32_t frame_length() { return frame_buf_length_; }
+
+  // Returns true if a new frame was detected and this frame is available for read.
+  bool frame_available() { return frame_available_; }
+
+#ifdef TEST_MODE
+  State get_state() { return state_; }
+#endif
+
+ private:
+  RxBuffer &rx_buffer_;
+  State state_{State::Lost};
+  bool frame_available_{false};
+  uint8_t frame_buffer_[FrameBufferLength] = {0};
+  uint32_t frame_buf_length_{0};
+  uint32_t marker_count_{0};
+
+  void restart_rx() {
+    marker_count_ = 0;
+    rx_buffer_.restart_rx(this);
+  }
+
+  bool data_received() { return rx_buffer_.received_length() > marker_count_; }
+  bool only_markers_received() { return rx_buffer_.received_length() == marker_count_; }
+  void process_received_data() {
+    // We strip incoming markers. First marker is stripped by restarting RX after frame start MARK,
+    // i.e. the first marker is not saved in RxBuffer; and now here we strip the last marker
+    // (which was saved to RxBuffer) by taking the received_length() - 1
+    frame_buf_length_ = rx_buffer_.received_length() - marker_count_;
+    memcpy(frame_buffer_, rx_buffer_.get() + marker_count_ - 1, frame_buf_length_);
```

Comment:
What happens here in case `frame_buf_length_ > FrameBufferLength` should we guard against that?